File: 2023/langs/php/1/solution2.php
<?php

$file_path = join("/", array(getcwd(), "..", "..", "..", "inputs", "1.txt"));
$contents = file_get_contents($file_path);
$calibration_data = explode("\n", $contents);
$numbers = str_split("1234567890");
$words = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
$tests = array_merge($numbers, $words);

$total = 0;

foreach ($calibration_data as $value) {
  $first_pos = strlen($value) + 20;
  $first = "";
  foreach ($tests as $test) {
    $pos = strpos($value, $test);
    if ($pos === false)
      continue;
    if ($first_pos < $pos)
      continue;
    $first_pos = $pos;
    $first = $test;
  }

  $last_pos = -1;
  $last = "";
  foreach ($tests as $test) {
    $pos = strrpos($value, $test);
    if ($pos === false)
      continue;
    if ($last_pos > $pos)
      continue;
    $last_pos = $pos;
    $last = $test;
  }

  if (in_array($first, $words, true))
    $first = array_search($first, $words, true) + 1;
  if (in_array($last, $words, true))
    $last = array_search($last, $words, true) + 1;

  $digits = $first . $last;
  $total += $digits;
}

print($total);

File: 2023/langs/python/1/solution2.py
import os

with open(os.path.join(os.getcwd(), "..", "..", "..", "inputs", "1.txt"), "r") as file:
    contents = file.read()

calibrationData = contents.split("\n")
numbers = list("1234567890")
words = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
tests = numbers + words

total = 0

for value in calibrationData:
    firstPos = len(value) + 20
    first = ""
    for test in tests:
        pos = value.find(test)
        if pos < 0:
            continue
        if firstPos < pos:
            continue
        firstPos = pos
        first = test

    lastPos = -1
    last = ""
    for test in tests:
        pos = value.rfind(test)
        if pos < 0:
            continue
        if lastPos > pos:
            continue
        lastPos = pos
        last = test

    if first in words:
        first = str(words.index(first) + 1)
    if last in words:
        last = str(words.index(last) + 1)

    digits = first + last
    total += int(digits)

print(total)
